File: azure-functions/csharp/Handlers/SubscriptionHandler.cs
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using System.Collections.Generic;
using Company.Sales.Api.Models;
using System;

namespace Company.Sales.Api.Handlers
{
    public static class SubscriptionHandler
    {
        internal const string CollectionName = "Subscriptions";

        [FunctionName("GetSubscriptions")]
        public static IActionResult GetSubscriptions(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "customers/{customerId}/subscriptions")] HttpRequest req,
            [CosmosDB(
                databaseName: "%DatabaseName%",
                collectionName: CollectionName,
                ConnectionStringSetting = "CosmosConnectionStr",
                SqlQuery = "SELECT * FROM s where s.customerId = {customerId}")] IEnumerable<Subscription> subscriptions,
            [CosmosDB(
                databaseName: "%DatabaseName%",
                collectionName: CustomerHandler.CollectionName,
                ConnectionStringSetting = "CosmosConnectionStr",
                Id = "{customerId}",
                PartitionKey = "{customerId}")] Customer customer,
            Guid customerId,
            ILogger log)
        {
            if (customer == null)
            {
                return new BadRequestObjectResult($"No customer with Id {customerId}");
            }
            else
            {
                return new OkObjectResult(subscriptions);
            }
        }

        [FunctionName("CreateSubscription")]
        public static async Task<IActionResult> CreateSubscription(
            [HttpTrigger(
                AuthorizationLevel.Anonymous,
                "post",
                Route = "customers/{customerId:guid}/subscriptions")] Subscription subscription,
            [CosmosDB(
                databaseName: "%DatabaseName%",
                collectionName: CustomerHandler.CollectionName,
                ConnectionStringSetting = "CosmosConnectionStr",
                Id = "{customerId}",
                PartitionKey = "{customerId}")] Customer customer,
            [CosmosDB(
                databaseName: "%DatabaseName%",
                collectionName: CollectionName,
                ConnectionStringSetting = "CosmosConnectionStr")] IAsyncCollector<Subscription> subscriptionCollector,
            Guid customerId,
            ILogger log)
        {
            if (customer == null)
            {
                return new BadRequestObjectResult($"No customer with Id {customerId}");
            }
            else if (subscription.Price <= 0M)
            {
                return new BadRequestObjectResult($"Price invalid.");
            }
            else
            {
                subscription.Id = Guid.NewGuid();
                subscription.CustomerId = customerId;
                subscription.DayOfMonth = DateTime.UtcNow.Day;

                await subscriptionCollector.AddAsync(subscription);

                return new CreatedResult($"/customers/{subscription.CustomerId}/subscription", subscription);
            }
        }

        [FunctionName("RetrieveDailySubscriptions")]
        public static void RetrieveDailySubscriptions(
            [TimerTrigger("0 0 0 * * *")]TimerInfo subscriptionTimer,
            [CosmosDB(
                databaseName: "%DatabaseName%",
                collectionName: CollectionName,
                ConnectionStringSetting = "CosmosConnectionStr",
                SqlQuery = "SELECT * FROM s where s.dayOfMonth = DateTimePart(\"d\", GetCurrentDateTime())")] IEnumerable<Subscription> subscriptions,
            [ServiceBus(
                "subscription-payments",
                Connection = "ServiceBusConnectionStr")] IAsyncCollector<Subscription> subscriptionCollector,
            ILogger log)
        {
            var tasks = new List<Task>();

            foreach (var subscription in subscriptions)
            {
                tasks.Add(subscriptionCollector.AddAsync(subscription));
            }

            Task.WaitAll(tasks.ToArray());
            log.LogInformation($"Processed {subscriptions.Count()} subscriptions.");
        }
    }
}

File: azure-functions/java/src/main/java/com/company/sales/api/handlers/SubscriptionHandler.java
package com.company.sales.api.handlers;

import com.microsoft.azure.functions.*;
import com.microsoft.azure.functions.annotation.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.UUID;

import com.company.sales.api.models.*;

public class SubscriptionHandler {
    final static String CollectionName = "Subscriptions";

    @FunctionName("GetSubscriptions")
    public HttpResponseMessage getSubscriptions(
            @HttpTrigger(
                name = "req",
                methods = {HttpMethod.GET},
                authLevel = AuthorizationLevel.ANONYMOUS,
                route = "customers/{customerId}/subscriptions") HttpRequestMessage<String> request,
            @CosmosDBInput(name = "subscriptionDatabase",
                databaseName = "%DatabaseName%",
                collectionName = CollectionName,
                sqlQuery = "select * from s where s.customerId = {customerId}",
                connectionStringSetting = "CosmosConnectionStr") Subscription[] subscriptions,
            @CosmosDBInput(name = "customerDatabase",
                databaseName = "%DatabaseName%",
                collectionName = CustomerHandler.CollectionName,
                id = "{customerId}",
                partitionKey = "{customerId}",
                connectionStringSetting = "CosmosConnectionStr") Customer customer,
            @BindingName("customerId") final UUID customerId,
            final ExecutionContext context) {
        if (customer == null) {
            return request.createResponseBuilder(HttpStatus.BAD_REQUEST).body("No customer with Id " + customerId).build();
        } else {
            return request.createResponseBuilder(HttpStatus.OK).body(subscriptions).build();
        }
    }

    @FunctionName("CreateSubscription")
    public HttpResponseMessage createSubscription(
            @HttpTrigger(
                name = "req",
                methods = {HttpMethod.POST},
                authLevel = AuthorizationLevel.ANONYMOUS,
                route = "customers/{customerId}/subscriptions") HttpRequestMessage<Subscription> request,
            @CosmosDBInput(name = "customerDatabase",
                databaseName = "%DatabaseName%",
                collectionName = CustomerHandler.CollectionName,
                id = "{customerId}",
                partitionKey = "{customerId}",
                connectionStringSetting = "CosmosConnectionStr") Customer customer,
            @CosmosDBOutput(name = "subscriptionDatabase",
                databaseName = "%DatabaseName%",
                collectionName = CollectionName,
                connectionStringSetting = "CosmosConnectionStr") OutputBinding<Subscription> subscriptionOutput,
            @BindingName("customerId") final UUID customerId,
            final ExecutionContext context) {
        Subscription subscription = request.getBody();

        if (customer == null) {
            return request.createResponseBuilder(HttpStatus.BAD_REQUEST).body("No customer with Id " + customerId).build();
        } else if (subscription.price.compareTo(BigDecimal.ZERO) <= 0) {
            return request.createResponseBuilder(HttpStatus.BAD_REQUEST).body("Price invalid.").build();
        } else {
            subscription.customerId = customerId;
            subscription.id = UUID.randomUUID();
            subscription.dayOfMonth = LocalDate.now().getDayOfMonth();
            subscriptionOutput.setValue(subscription);

            return request.createResponseBuilder(HttpStatus.CREATED).body(subscription).build();
        }
    }

    @FunctionName("RetrieveDailySubscriptions")
    public void retrieveDailySubscriptions(
        @TimerTrigger(
            name = "timer",
            schedule = "0 0 0 * * *") String timerInfo,
        @CosmosDBInput(name = "subscriptionDatabase",
            databaseName = "%DatabaseName%",
            collectionName = CollectionName,
            sqlQuery = "SELECT * FROM s where s.dayOfMonth = DateTimePart(\"d\", GetCurrentDateTime())",
            connectionStringSetting = "CosmosConnectionStr") Subscription[] subscriptions,
        @ServiceBusQueueOutput(name = "sb",
            queueName = "%PaymentQueue%",
            connection = "ServiceBusConnectionStr") OutputBinding<Subscription[]> subscriptionOutput,
        final ExecutionContext context) {
        subscriptionOutput.setValue(subscriptions);
    }
}
